### PR TITLE
Fix for missing nested namespace identifier while declaring variables

### LINKED ISSUE #521: Nested Namespace specifier missing in clad generated derivative code
I found an issue with missing nested namespace in derivative code generated by clad as follow:

For the following function:
```c++
double test(double x, double y){
    return std::sin(x)+ std::cos(y);
}
```
Clad generated below derivative:
```C++
double test_darg0(double x, double y) {
    double _d_x = 1;
    double _d_y = 0;
    ValueAndPushforward<double, double> _t0 = clad::custom_derivatives::sin_pushforward(x, _d_x);
    ValueAndPushforward<double, double> _t1 = clad::custom_derivatives::cos_pushforward(y, _d_y);
    return _t0.pushforward + _t1.pushforward;
}
```
When using this derivative function directly in a program, the following errors were produced:

```
clad/demos/namespaceIssue.cpp:10:5: error: no template named 'ValueAndPushforward'; did you mean 'clad::ValueAndPushforward'?
    ValueAndPushforward<double, double> _t0 = clad::custom_derivatives::sin_pushforward(x, _d_x);
    ^~~~~~~~~~~~~~~~~~~
    clad::ValueAndPushforward
/Users/vidushi/development/use-clad/clad/test/Hessian/../../include/clad/Differentiator/BuiltinDerivatives.h:19:42: note: 'clad::ValueAndPushforward' declared here
template <typename T, typename U> struct ValueAndPushforward {
                                         ^
clad/demos/namespaceIssue.cpp:11:41: error: use of undeclared identifier '_t1'; did you mean '_t0'?
    ValueAndPushforward<double, double> _t1 = clad::custom_derivatives::cos_pushforward(y, _d_y);
                                        ^~~
                                        _t0
clad/demos/namespaceIssue.cpp:10:41: note: '_t0' declared here
    ValueAndPushforward<double, double> _t0 = clad::custom_derivatives::sin_pushforward(x, _d_x);
                                        ^
clad/demos/namespaceIssue.cpp:11:40: error: expected ';' after expression
    ValueAndPushforward<double, double> _t1 = clad::custom_derivatives::cos_pushforward(y, _d_y);
                                       ^
                                       ;
clad/demos/namespaceIssue.cpp:11:5: error: use of undeclared identifier 'ValueAndPushforward'
    ValueAndPushforward<double, double> _t1 = clad::custom_derivatives::cos_pushforward(y, _d_y);
    ^
clad/demos/namespaceIssue.cpp:12:30: error: use of undeclared identifier '_t1'; did you mean '_t0'?
    return _t0.pushforward + _t1.pushforward;
                             ^~~
                             _t0
clad/demos/namespaceIssue.cpp:10:41: note: '_t0' declared here
    ValueAndPushforward<double, double> _t0 = clad::custom_derivatives::sin_pushforward(x, _d_x);
                                        ^
5 errors generated.
```

If we change the `ValueAndPushforward` to `clad::ValueAndPushforward` in the derivative function, the program runs without error and produces the correct output.

### PR DESCRIPTION
**Pull request by vaithak**
Add the missing nested namespace identifier from the code generated in forward mode using custom derivatives.

closes #521

### CODE REVIEW

File: lib/Differentiator/ForwardModeVisitor.cpp [hunk 1]
```diff
@@ -1298,8 +1298,15 @@ namespace clad {
...
       return StmtDiff(callDiff, nullptr);
-    auto valueAndPushforward =
-        StoreAndRef(callDiff, "_t", /*forceDeclCreation=*/true);
+
+    // get return type of pushforward function.
+    CXXScopeSpec CSS;
+    CSS.Extend(m_Context, GetCladNamespace(), noLoc, noLoc);
+    NestedNameSpecifier* NS = CSS.getScopeRep();
+    QualType returnType = m_Context.getElaboratedType(ETK_None, NS, callDiff->getType());
+    auto valueAndPushforward = StoreAndRef(callDiff, returnType, getCurrentBlock(), "_t",
```
Comment (vgvassilev) [inline lib/Differentiator/ForwardModeVisitor.cpp:1307]: Would it make sense to lift this code in `StoreAndRef`?

**Comment by parth-07** [comment]
Hi @vaithak 
Thank you for working on this. The solution used in this pull request does not solve the problem at its core. Some of the issues which I found in this pull request are:
1. It adds `clad::` nested namespace specifier without performing any initial checks. For example, what if
`callDiff->getType()` already has `clad::` nested namespace specifier? As a concrete example:

```cpp
double mul(double i, double j) {
  return i * j;
}

double fn(double i, double j) {
  return mul(i, j);
}
```

Differentiating `fn` w.r.t `i` with your change will generate the following derivative function:

```cpp
double fn_darg0(double i, double j) {
    double _d_i = 1;
    double _d_j = 0;
    clad::clad::ValueAndPushforward<double, double> _t0 = mul_pushforward(i, j, _d_i, _d_j);
    return _t0.pushforward;
}
```
Please note the incorrect nested namespace specifier in the generated function: `clad::clad::ValueAndPushforward`.

2. It doesn't solve the problem at its core. The issue at the core can (informally) be stated as follows: declaration references in the generated function have missing nested name specifiers. Your solution only tries to fix the issue for just one specific case: `ValueAndPushforward` object declarations. As a concrete example:

```cpp
namespace A {
    namespace B {
        namespace C {
            double add (double i, double j) {
                return i + j;
            }
        }
    }
}

double fn(double i, double j) {
    return A::B::C::add(i, j);
}
```

Differentiating `fn` w.r.t `i` using forward mode AD generates the following derivative function:

```cpp
double fn_darg0(double i, double j) {
    double _d_i = 1;
    double _d_j = 0;
    clad::ValueAndPushforward<double, double> _t0 = add_pushforward(i, j, _d_i, _d_j);
    return _t0.pushforward;
}
```

Please note the missing nested name specifiers in the `add_pushforward` declaration reference.
-------------------------
It will be best to solve this problem at it's core by adding nested namespace specifiers to declaration references at key checkpoints in the codebase. To build nested namespace specifier in a general way you can use the function, `clad::utils::BuildNNS`.

**Comment by vgvassilev** [comment]
@vaithak, ping. Also I was wondering if you are pursuing GSoC this year, if so feel free to send me an email.

**Comment by vaithak** [comment]
Hi, I have modified my code to add the namespace generically during variable declaration. 
To avoid double adding of namespaces as @parth-07 mentioned, I have used `Type.getDesugaredType` for now, but it is problematic as it won't work nicely with user-defined types, for ex. it generates `tape_impl` in code instead of `tape`.

Is there a function to just remove namespace identifiers from `clang::QualType` ?

File: lib/Differentiator/ForwardModeVisitor.cpp (hunk 1 above)
Comment (vaithak) [reply lib/Differentiator/ForwardModeVisitor.cpp]: added it generically during variable declaration but need some help as mentioned in https://github.com/vgvassilev/clad/pull/524#issuecomment-1470916617.

**Comment by parth-07** [comment]
> Hi, I have modified my code to add the namespace generically during variable declaration.

That's great. Thank you for working on this.

> * To avoid double adding of namespaces as @parth-07 mentioned, I have used `Type.getDesugaredType` for now, but it is problematic as it won't work nicely with user-defined types, for ex. it generates `tape_impl` in code instead of `tape`.
>   
>   * Is there a function to just remove namespace identifiers from `clang::QualType` ?

We don't have an already existing function for this. I will look into it, and will update you by EOD tomorrow.

> * Regarding the second point, in this PR I am aiming to just focus on namespace issues during variable declarations, for generated functions we can have a separate issue. What do you suggest?

Yes, this sounds perfect.

**Comment by parth-07** [comment]
@vaithak Here's a simplistic version of a function that can be used to remove all namespace identifiers from a type.

```cpp
clang::QualType getUnqualifiedType(clang::QualType T) {
  if (auto elaboratedType = llvm::dyn_cast<clang::ElaboratedType>(T)) {
    auto tagDecl = elaboratedType->getAsTagDecl();
    auto nonElaboratedType = tagDecl->getTypeForDecl()->getCanonicalTypeInternal();
    return nonElaboratedType;
  }
  return T;
}

**Comment by vaithak** [comment]
Hi @parth-07, thanks for replying. 
The function that you suggested is also not working with custom defined types like `tape` or `TensorD5` (Tensor<double, 5>.

**Comment by parth-07** [comment]
Hi @vaithak 

> The function that you suggested is also not working with custom-defined types like tape or TensorD5

Thank you for the update. I'll look into it.

**Comment by vgvassilev** [comment]
> Hi @vaithak
> 
> > The function that you suggested is also not working with custom-defined types like tape or TensorD5
> 
> Thank you for the update. I'll look into it.

@parth-07 ping.

**Comment by parth-07** [comment]
> @parth-07 ping.

Thank you for the ping. I will soon look into it, and update here.

**Comment by parth-07** [comment]
Hi @vaithak, 

It seems that typedef names have correct namespace specifiers already. For now, we can change the below code:
```cpp
 if (typePtr->isRecordType()) {
   // ...
 }
```
to:
```cpp
 if (typePtr->isRecordType() && !Type->isTypedefNameType()) {
   // ...
 }
```

File: lib/Differentiator/VisitorBase.cpp [hunk 2]
```diff
@@ -63,6 +64,17 @@ namespace clad {
                                      TypeSourceInfo* TSI,
                                      VarDecl::InitializationStyle IS) {
 
+    // add namespace specifier in variable declaration.
+    auto typePtr = Type.getTypePtr();
```
Comment (parth-07) [inline lib/Differentiator/VisitorBase.cpp:68]: We should perhaps move this code to a separate function.

**Review by parth-07 — COMMENTED** [review]
The patch looks good. It seems `clang::Type::isTypedefNameType` function was added in LLVM 12. You would need to make this function compatible with Clang 8 - Clang 11. We typically store such functions in `include/clad/Differentiator/Compatibility.h`. Also, tests need to be fixed before this can be merged.

File: lib/Differentiator/VisitorBase.cpp (hunk 2 above)
Comment (vaithak) [reply lib/Differentiator/VisitorBase.cpp]: moved to clad utils

**Comment by vgvassilev** [comment]
Can you squash they changes into one?

**Review by parth-07 — APPROVED** [review]
Looks good. Please squash the commits into one commit as @vgvassilev suggested.

File: test/ForwardMode/constexprTest.C [hunk 3]
```diff
@@ -1,43 +0,0 @@
-// RUN: %cladclang -lm -lstdc++ %s -I%S/../../include -oconstexprTest.out | FileCheck %s
```
Comment (vgvassilev) [inline test/ForwardMode/constexprTest.C:1]: Why did we delete this test?

**Comment by vaithak** [comment]
@vgvassilev @parth-07 does this require any other changes?

**Review by vgvassilev — APPROVED** [review]

**Pull request merged**